File: leetcode/search/1428. Leftmost Column with at Least a One/py/main.py
# """
# This is BinaryMatrix's API interface.
# You should not implement it, or speculate about its implementation
# """
#class BinaryMatrix(object):
#    def get(self, row: int, col: int) -> int:
#    def dimensions(self) -> list[]:

#  Start at Top Right, Move Only Left and Down
# Time O(N + M)
# Space O(1)
class Solution:
    def leftMostColumnWithOne(self, binaryMatrix: 'BinaryMatrix') -> int:
        rows, cols = binaryMatrix.dimensions()
        
        current_row = 0
        current_col = cols - 1
        
        while current_row < rows and current_col >= 0:
            if binaryMatrix.get(current_row, current_col) == 0:
                current_row += 1
            else:
                current_col -= 1
        
        return -1 if current_col == cols - 1 else current_col + 1
        

File: leetcode/search/1428. Leftmost Column with at Least a One/js/main.js
// Time O(N * LogM)
// Space O(1)
const leftMostColumnWithOne = binaryMatrix => {
  let [rows, cols] = binaryMatrix.dimensions();
  let INF = Number.MAX_VALUE;
  let res = INF;

  for (let i = 0; i < rows; i++) {
    let index = search(i, cols);

    if (index != -1) {
      res = Math.min(res, index);
    }
  }

  return res == INF ? -1 : res;

  function search(row, cols) {
    let lo = 0;
    let hi = cols - 1;

    while (lo < hi) {
      let mid = lo + Math.floor((hi - lo) / 2);

      if (binaryMatrix.get(row, mid) == 0) {
        lo = mid + 1;
      } else {
        hi = mid;
        ans = hi;
      }
    }

    return binaryMatrix.get(row, lo) == 1 ? lo : -1;
  }
};
